hazelcast/src/main/java/com/hazelcast/map/impl/operation/MergeOperation.java

```diff
@@ -118,6 +113,23 @@ private void merge(MapMergeTypes mergingEntry) {
         }
     }
 
+    private void publishListenerEvent(Data dataKey, Data oldValue, Data dataValue) {
+        if (hasMapListener) {
+            mapEventPublisher.publishEvent(getCallerAddress(), name, MERGED, dataKey, oldValue, dataValue);
+        }
+    }
+
+    private void publishWanReplication(Data dataKey, Data dataValue) {
+        if (hasWanReplication) {
+            if (dataValue == null) {
+                mapEventPublisher.publishWanReplicationRemove(name, dataKey, Clock.currentTimeMillis());
+            } else {
+                EntryView<Data, Data> entryView = createSimpleEntryView(dataKey, dataValue, recordStore.getRecord(dataKey));
+                mapEventPublisher.publishWanReplicationUpdate(name, entryView);
+            }
+        }
+    }
+
     private Data getValueOrPostProcessedValue(Data dataKey, Data dataValue) {
```

Comment:
Is it possible for the `getValueOrPostProcessedValue` to throw a NPE if the record store is post processing and the record has been removed?

Reply:
I have no idea what this post processing is.

Reply:
Hmm, it looks to me like the method only makes sense, if you already have a `dataValue` (e.g. as return value from the called container method). And in case of post-processing (whatever this is exactly), you have to fetch the value again (to get the processed one). But we fetch the value anyway from the `recordStore` and give it as parameter to `getValueOrPostProcessedValue()`, so the whole method has no purpose here.

And yes, you should be right about the NPE.